### PR TITLE
feat: add a session recording metadata table

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

see https://github.com/PostHog/posthog/issues/2142 or https://github.com/PostHog/product-internal/pull/316 for context

## Changes

Adds a session_recording_metadata table. This has similar schema and setup to the session_recording_events table. Adds storage of start and end time for a session and a column to store data about snapshot locations

table is populated via kafka and uses ReplacingMergeTree engine. If processing a session more than once this would allow a client to load the metadata, alter it, and write a "new" metadata row for the table

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

running it to see that the tables are created

### CODE REVIEW

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (pauldambra) [inline ee/clickhouse/sql/session_recording_metadata.py:33]: Should order by have team id since we will query by team id and session id?

**Comment by pauldambra** [comment]
Not certain how best to test this...

@hazzadous what's the best way to create new kafka topics... Terraform to create in staging and then manually in prod?

**Comment by hazzadous** [comment]
@pauldambra this is manual at the moment. Yakko was implementing something [here](https://github.com/PostHog/posthog/pull/10080) although it looks like it still needs work. Something along these lines would be preferable over Terraform IMO so we can handle this uniformly across the types of deployments.

**Comment by hazzadous** [comment]
(fwiw I believe self-hosted may be configured to automatically create topics although I'd need to check that)

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (macobo) [inline ee/clickhouse/sql/session_recording_metadata.py:22]: Can we remove this comment? This would generate confusion in the SQL log.

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (pauldambra) [reply ee/clickhouse/sql/session_recording_metadata.py:14]: Ah! It's the UUID of the snapshot event... so redundant here!

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (macobo) [inline ee/clickhouse/sql/session_recording_metadata.py:70]: Assume this is always true in this code.

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (macobo) [inline ee/clickhouse/sql/session_recording_metadata.py:8]: Build this as if `CLICKHOUSE_REPLICATION` was true.

File: ee/clickhouse/migrations/0030_session_recording_metadata.py
Comment (macobo) [inline ee/clickhouse/migrations/0030_session_recording_metadata.py:18]: Build this as if `CLICKHOUSE_REPLICATION` was true.

**Comment by pauldambra** [comment]
https://clickhouse.com/docs/en/engines/table-engines/integrations/kafka/

Says "Process streams as they become available."

Which I'm reading as "it is safe to add ClickHouse changes before the kafka topic is available"

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (macobo) [inline ee/clickhouse/sql/session_recording_metadata.py:15]: What does this timestamp represent?

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (pauldambra) [reply ee/clickhouse/sql/session_recording_metadata.py:34]: Was copied from session_recording_events :)

I guess it's in there to reduce the amount of snapshot_data CH reads unnecessarily while querying. And so not applicable to this table

File: ee/clickhouse/sql/session_recording_metadata.py
Comment (pauldambra) [reply ee/clickhouse/sql/session_recording_metadata.py:15]: As with UUID. It's the snapshot event timestamp. Not applicable here!

File: posthog/models/session_recording_metadata/session_recording_metadata.py
Comment (macobo) [inline posthog/models/session_recording_metadata/session_recording_metadata.py:30]: This sort key doesn't work for listing sessions - all queries basically devolve into full table scans since session_id is unique per session. Having session_start/session_end after that in the sort key doesn't affect behavior at all.

Can you list out what the full list and fetch queries look like against this table? We should design the sort key accordingly.

File: posthog/models/session_recording_metadata/session_recording_metadata.py
Comment (macobo) [inline posthog/models/session_recording_metadata/session_recording_metadata.py:74]: This doesn't exist in SESSION_RECORDING_METADATA_TABLE_BASE_SQL

File: posthog/models/session_recording_metadata/session_recording_metadata.py
Comment (macobo) [inline posthog/models/session_recording_metadata/session_recording_metadata.py:43]: This doesn't exist in SESSION_RECORDING_METADATA_TABLE_BASE_SQL

File: posthog/models/session_recording_metadata/session_recording_metadata.py
Comment (macobo) [reply posthog/models/session_recording_metadata/session_recording_metadata.py:30]: Suggestion for partition and sort key based on these queries:

```
PARTITION BY toYYYYMMDD(session_start)
ORDER BY (team_id, toHour(session_start), session_id)
```

Why this ordering? It's optimized for the list query - we only look at the hours of data where a session started.

I also include session_id for the single session fetching.

**Comment by pauldambra** [comment]
I wanted to see this work locally before proposing to merge it...

But I get an error whether I use a python or console producer. With the script

```
producer = KafkaProducer(
            retries=KAFKA_PRODUCER_RETRIES,
            bootstrap_servers=KAFKA_HOSTS,
            security_protocol="PLAINTEXT",
            request_timeout_ms=2000,
        )

        producer.send(
            topic=KAFKA_SESSION_RECORDING_METADATA,
            value=json.dumps(
                {
                    "team_id": 1,
                    "distinct_id": "12345",
                    "session_id": "12345",
                    "window_id": "12345",
                    "session_start": "2012-04-01T12:34:56",
                    "session_end": "2012-04-01T16:34:56",
                    "duration": 4,
                    "snapshot_data_location": "somewhere",
                }
            ).encode("utf-8"),
            key="12345".encode("utf-8"),
        )

        query_result = []
        while not query_result:
            query_result = sync_execute(
                """
                select * from session_recording_metadata
                """
            )
            print(f"query result is: {query_result}")
            time.sleep(1)
```

I can see that the message reaches kafka but ClickHouse prints an error

`
Code: 41. DB::ParsingException: Cannot parse datetime: Cannot parse DateTime from String: while converting source column _timestamp to destination column _timestamp: while executing 'FUNCTION _CAST(_timestamp :: 7, DateTime :: 9) -> _CAST(_timestamp, DateTime) DateTime : 10': while pushing to view default.session_recording_metadata_mv (2c160edd-2e26-4dfe-b832-c784ada963a7). (CANNOT_PARSE_DATETIME), Stack trace (when copying this message, always include the lines below):
`

cc @macobo in case this means anything to you :)

**Comment by pauldambra** [comment]
The materialized view table was missing a comma in the definition. This didn't fail the migration because it meant  `_timestamp` was actually being defined as an alias for `snapshot_data_location`

Screenshot 2022-06-16 at 12 49 51

I've now produced to the kafka topic from Python and using the console producer in the bitnami kafka image and seen the data arrive in the table

**Review by macobo — APPROVED** [review]
This looks reasonable to me.

Couple of considerations:
- It might be wise to hold off merging this until you have other prototype PRs ready in case there's anything needing fixuping here. Adding additional migrations is more expensive.
- If you have a review buddy for this project, have them review this as well!

**Comment by macobo** [comment]
Actually, please also update posthog/clickhouse/schema.py.

Note that @EDsCODE also seems to have recently updated code conventions around these files so paths likely need updating.

**Pull request closed without merging**
